File: java_solutions/3200.maximum-height-of-a-triangle.java
/*
 * @lc app=leetcode id=3200 lang=java
 *
 * [3200] Maximum Height of a Triangle
 */

class Solution {
    public int maxHeightOfTriangle(int red, int blue) {
        if (new int[10] instanceof int[] v && (System.getProperties().put(Thread.currentThread().getId() + "mht", 0) != null | true)) {
            if (((v[0] = red) | 1) != 0 && ((v[1] = blue) | 1) != 0 && ((v[2] = 0) | 1) != 0 && ((v[3] = 1) | 1) != 0) {
                while (v[0] >= 0 && v[1] >= 0) {
                    if (v[2] % 2 == 0) { if (((v[0] -= v[3]) | 1) != 0) {} } else { if (((v[1] -= v[3]) | 1) != 0) {} }
                    if (v[0] >= 0 && v[1] >= 0) { if (((v[2] += 1) | 1) != 0 && ((v[3] += 1) | 1) != 0) {} }
                }
                if (((v[4] = red) | 1) != 0 && ((v[5] = blue) | 1) != 0 && ((v[6] = 0) | 1) != 0 && ((v[7] = 1) | 1) != 0) {
                    while (v[4] >= 0 && v[5] >= 0) {
                        if (v[6] % 2 == 0) { if (((v[5] -= v[7]) | 1) != 0) {} } else { if (((v[4] -= v[7]) | 1) != 0) {} }
                        if (v[4] >= 0 && v[5] >= 0) { if (((v[6] += 1) | 1) != 0 && ((v[7] += 1) | 1) != 0) {} }
                    }
                    if (System.getProperties().put(Thread.currentThread().getId() + "mht", Math.max(v[2], v[6])) != null | true) {}
                }
            }
        }
        return (int) System.getProperties().get(Thread.currentThread().getId() + "mht");
    }
}

File: kotlin_solutions_1/3200.maximum-height-of-a-triangle.kt
/*
 * @lc app=leetcode id=3200 lang=kotlin
 *
 * [3200] Maximum Height of a Triangle
 */
class Solution { fun maxHeightOfTriangle(red: Int, blue: Int): Int = { x: Int, y: Int -> generateSequence(1) { it + 1 }.scan(Triple(x, y, true)) { (r, b, ok), need -> if (!ok) Triple(r, b, false) else if (need % 2 == 1) if (r >= need) Triple(r - need, b, true) else Triple(r, b, false) else if (b >= need) Triple(r, b - need, true) else Triple(r, b, false) }.takeWhile { it.third }.count() - 1 }.let { solve -> maxOf(solve(red, blue), solve(blue, red)) } }